File: ladders/cses/combinations.cpp
#include <iostream>
#include <vector>

using namespace std;

void combinations(const vector<int> &a, int curr_idx, int n,
                  vector<int> curr_comb, vector<vector<int>> &combs, int k) {
  if (curr_comb.size() == k) {
    combs.push_back(curr_comb);
    return;
  }

  if (curr_idx == n) {
    return;
  }

  curr_comb.push_back(a[curr_idx]);
  combinations(a, curr_idx + 1, n, curr_comb, combs, k);
  curr_comb.pop_back();
  combinations(a, curr_idx + 1, n, curr_comb, combs, k);
}

vector<vector<int>> combinations(const vector<int> &a, int n, int k) {
  vector<vector<int>> combs;
  vector<int> curr_comb;
  combinations(a, 0, n, curr_comb, combs, k);

  return combs;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, k;
  cin >> n >> k;

  vector<int> a(n);
  for (int i = 0; i < n; i++) {
    cin >> a[i];
  }

  auto combs = combinations(a, n, k);
  for (auto comb : combs) {
    for (int c : comb) {
      cout << c << " ";
    }
    cout << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/main_iterations/iteration5/Combinations.java
import java.util.*;

public class Combinations {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int k = in.nextInt();

      if (k < 0 || k > n) {
        throw new IllegalArgumentException("invalid value of k");
      }

      int[] a = new int[n];
      for (int i = 0; i < n; i++) {
        a[i] = in.nextInt();
      }

      List<List<Integer>> combinations = combinations(a, n, k);
      for (List<Integer> comb : combinations) {
        for (int c : comb) {
          System.out.printf("%d ", c);
        }
        System.out.println();
      }
    }
  }

  public static List<List<Integer>> combinations(int[] a, int n, int k) {
    List<List<Integer>> combs = new ArrayList<>();
    combinations(a, 0, n, new ArrayList<>(), combs, k);

    return combs;
  }

  private static void combinations(int[] a, int currIdx, int n,
                                   List<Integer> currComb,
                                   List<List<Integer>> combs, int k) {
    if (currComb.size() == k) {
      List<Integer> tmp = new ArrayList<>();
      for (int c : currComb) {
        tmp.add(c);
      }
      combs.add(tmp);
      return;
    }

    if (currIdx == n) {
      return;
    }
    currComb.add(a[currIdx]);
    combinations(a, currIdx + 1, n, currComb, combs, k);
    currComb.remove(currComb.size() - 1);
    combinations(a, currIdx + 1, n, currComb, combs, k);
  }
}